In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from pydataset import data

# 1. Attendance Data
Load the `attendance.csv` file and calculate an attendance percentage for each student. 
- One half day is worth 50% of a full day, and 
- 10 tardies is equal to one absence.

In [2]:
attendance = pd.read_csv('untidy-data/attendance.csv', encoding = 'unicode escape')

In [143]:
attendance

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [4]:
#melted the multiple columns into one col:date and product into another col:attendance
attendance_melt = attendance.melt(id_vars = 'Unnamed: 0',
                                 var_name = 'date', value_name = 'attendance')
attendance_melt

Unnamed: 0        date attendance
0       Sally  2018-01-01          P
1        Jane  2018-01-01          A
2       Billy  2018-01-01          A
3        John  2018-01-01          P
4       Sally  2018-01-02          T
5        Jane  2018-01-02          P
6       Billy  2018-01-02          T
7        John  2018-01-02          T
8       Sally  2018-01-03          T
9        Jane  2018-01-03          T
10      Billy  2018-01-03          A
11       John  2018-01-03          H
12      Sally  2018-01-04          H
13       Jane  2018-01-04          T
14      Billy  2018-01-04          A
15       John  2018-01-04          P
16      Sally  2018-01-05          P
17       Jane  2018-01-05          T
18      Billy  2018-01-05          H
19       John  2018-01-05          P
20      Sally  2018-01-06          A
21       Jane  2018-01-06          T
22      Billy  2018-01-06          T
23       John  2018-01-06          T
24      Sally  2018-01-07          T
25       Jane  2018-01-07          A
26      Billy  2018-01-07          P
27       John  2018-01-07          P
28      Sally  2018-01-08          T
29       Jane  2018-01-08          T
30      Billy  2018-01-08          T
31       John  2018-01-08          P

In [5]:
#renamed first col to a suitable name
attendance_melt.rename(columns = {'Unnamed: 0' : 'name'}, inplace = True)
attendance_melt.head()

name        date attendance
0  Sally  2018-01-01          P
1   Jane  2018-01-01          A
2  Billy  2018-01-01          A
3   John  2018-01-01          P
4  Sally  2018-01-02          T

In [139]:
#attendance_melt.drop(columns = 'date', inplace = True)
#attendance_melt

In [140]:
# replaced attendance variable to numerical values
# Present=1, Absent=0, Halfday=0.5, Tardy=.9

#attendance_melt.replace((to_replace ="P", value = 1), (to_replace = 'T', value =  0.9))
attendance_melt = attendance_melt.replace(to_replace ="P", value = 1).replace(to_replace ="T", value = 0.9).replace(to_replace ="A", value = 0).replace(to_replace ="H", value = 0.5)
attendance_melt.head(12)

name        date  attendance
0   Sally  2018-01-01         1.0
1    Jane  2018-01-01         0.0
2   Billy  2018-01-01         0.0
3    John  2018-01-01         1.0
4   Sally  2018-01-02         0.9
5    Jane  2018-01-02         1.0
6   Billy  2018-01-02         0.9
7    John  2018-01-02         0.9
8   Sally  2018-01-03         0.9
9    Jane  2018-01-03         0.9
10  Billy  2018-01-03         0.0
11   John  2018-01-03         0.5

In [25]:
#created an avg of each student's attendance over the 8 days
attendance_melt.groupby('name').mean('attendance')

attendance
name             
Billy      0.5250
Jane       0.6875
John       0.9125
Sally      0.7625

In [141]:
#attendance_tidy = attendance_melt.pivot_table(index= 'date', 
#                                              columns = 'name', values = 'attendance')

In [142]:
#attendance_tidy

In [11]:
#attendance_tidy.shape

(8, 4)

In [162]:
#attendance_tidy.T

#### A different take with same results

In [144]:
attendance.rename(columns = {'Unnamed: 0' : 'name'}, inplace = True)
attendance.head()

name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0  Sally          P          T          T          H          P          A   
1   Jane          A          P          T          T          T          T   
2  Billy          A          T          A          A          H          T   
3   John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [146]:
att = attendance.replace({'P': 1.0, 'A': 0.0, 'T': 0.9, 'H': 0.5})
att

name  2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  \
0  Sally         1.0         0.9         0.9         0.5         1.0   
1   Jane         0.0         1.0         0.9         0.9         0.9   
2  Billy         0.0         0.9         0.0         0.0         0.5   
3   John         1.0         0.9         0.5         1.0         1.0   

   2018-01-06  2018-01-07  2018-01-08  
0         0.0         0.9         0.9  
1         0.9         0.0         0.9  
2         0.9         1.0         0.9  
3         0.9         1.0         1.0

In [160]:
# changed 'name' col to be my index
att_tidy = att.pivot_table(index = 'name')
att_tidy

2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  2018-01-06  \
name                                                                            
Billy         0.0         0.9         0.0         0.0         0.5         0.9   
Jane          0.0         1.0         0.9         0.9         0.9         0.9   
John          1.0         0.9         0.5         1.0         1.0         0.9   
Sally         1.0         0.9         0.9         0.5         1.0         0.0   

       2018-01-07  2018-01-08  
name                           
Billy         1.0         0.9  
Jane          0.0         0.9  
John          1.0         1.0  
Sally         0.9         0.9

In [161]:
att_tidy.mean(axis=1)


name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
dtype: float64

# 2. Coffee Levels

2a. Read the `coffee_levels.csv` file.

In [26]:
coffee_levels = pd.read_csv('untidy-data/coffee_levels.csv')
coffee_levels.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

In [27]:
coffee_levels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   hour           30 non-null     int64  
 1   coffee_carafe  30 non-null     object 
 2   coffee_amount  30 non-null     float64
dtypes: float64(1), int64(1), object(1)
memory usage: 848.0+ bytes


In [29]:
coffee_levels

hour coffee_carafe  coffee_amount
0      8             x       0.816164
1      9             x       0.451018
2     10             x       0.843279
3     11             x       0.335533
4     12             x       0.898291
5     13             x       0.310711
6     14             x       0.507288
7     15             x       0.215043
8     16             x       0.183891
9     17             x       0.391560
10     8             y       0.189297
11     9             y       0.521502
12    10             y       0.023163
13    11             y       0.235529
14    12             y       0.017009
15    13             y       0.997464
16    14             y       0.058361
17    15             y       0.144644
18    16             y       0.544676
19    17             y       0.594126
20     8             z       0.999264
21     9             z       0.915990
22    10             z       0.144928
23    11             z       0.311495
24    12             z       0.771947
25    13             z       0.398520
26    14             z       0.864464
27    15             z       0.436364
28    16             z       0.280621
29    17             z       0.436677

2b. Transform the data so that each carafe is in it's own **column**.

In [30]:
coffee_levels.head(2)

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018

In [163]:
coffee_pivot = coffee_levels.pivot_table(index = 'hour', 
                         columns = 'coffee_carafe', values = 'coffee_amount')
coffee_pivot

coffee_carafe         x         y         z
hour                                       
8              0.816164  0.189297  0.999264
9              0.451018  0.521502  0.915990
10             0.843279  0.023163  0.144928
11             0.335533  0.235529  0.311495
12             0.898291  0.017009  0.771947
13             0.310711  0.997464  0.398520
14             0.507288  0.058361  0.864464
15             0.215043  0.144644  0.436364
16             0.183891  0.544676  0.280621
17             0.391560  0.594126  0.436677

2c. Is this the best shape for the data?

In [164]:
# I think this this APPEARS to be neater, although it is aggregating/averaging data
# It would be best to maybe include a 'day' index to show ALL of the data.

# 3. Cake Recipes

3a. Read the `cake_recipes.csv` data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [33]:
cake_recipes = pd.read_csv('untidy-data/cake_recipes.csv')
cake_recipes.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [34]:
cake_recipes.shape

(8, 5)

In [35]:
cake_recipes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   recipe:position  8 non-null      object 
 1   225              8 non-null      float64
 2   250              8 non-null      float64
 3   275              8 non-null      float64
 4   300              8 non-null      float64
dtypes: float64(4), object(1)
memory usage: 448.0+ bytes


In [36]:
cake_recipes

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

3b. Tidy the data as necessary.

In [42]:
#rename 1st col to remove ':'
cake_recipes = cake_recipes.rename(columns = {'recipe:position' : 'recipe_position'})

In [43]:
cake_recipes.head()

recipe_position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [44]:
#create 2 new cols by splitting col[0]
cake_recipes[['recipe_name', 'oven_position']] = cake_recipes.recipe_position.str.split(':', expand = True)

In [46]:
cake_recipes.head()

recipe_position        225        250        275        300 recipe_name  \
0        a:bottom  61.738655  53.912627  74.414730  98.786784           a   
1           a:top  51.709751  52.009735  68.576858  50.228470           a   
2        b:bottom  57.095320  61.904369  61.196980  99.248541           b   
3           b:top  82.455004  95.224151  98.594881  58.169349           b   
4        c:bottom  96.470207  52.001358  92.893227  65.473084           c   

  oven_position  
0        bottom  
1           top  
2        bottom  
3           top  
4        bottom

In [47]:
#removing redundant col[0]
cake_recipes.drop(columns = 'recipe_position', inplace=True)

In [48]:
cake_recipes.head()

225        250        275        300 recipe_name oven_position
0  61.738655  53.912627  74.414730  98.786784           a        bottom
1  51.709751  52.009735  68.576858  50.228470           a           top
2  57.095320  61.904369  61.196980  99.248541           b        bottom
3  82.455004  95.224151  98.594881  58.169349           b           top
4  96.470207  52.001358  92.893227  65.473084           c        bottom

In [57]:
#melting 'temp' columns into 1, and creating 'accuracy' col with the product
cake_melt = cake_recipes.melt(id_vars = ['recipe_name', 'oven_position'], 
                             var_name = 'temp', value_name = 'accuracy')

In [58]:
cake_melt.head()

recipe_name oven_position temp   accuracy
0           a        bottom  225  61.738655
1           a           top  225  51.709751
2           b        bottom  225  57.095320
3           b           top  225  82.455004
4           c        bottom  225  96.470207

In [93]:
#avging accuracy of ea recipe
cake_melt.groupby('recipe_name').mean()

accuracy
recipe_name           
a            63.922201
b            76.736074
c            75.874748
d            62.864844

In [59]:
#cake_tidy = cake_melt.pivot_table(index = ['recipe_name', 'oven_position'], 
                                 columns = 'temp', values = 'accuracy')

In [60]:
#cake_tidy

temp                             225        250        275        300
recipe_name oven_position                                            
a           bottom         61.738655  53.912627  74.414730  98.786784
            top            51.709751  52.009735  68.576858  50.228470
b           bottom         57.095320  61.904369  61.196980  99.248541
            top            82.455004  95.224151  98.594881  58.169349
c           bottom         96.470207  52.001358  92.893227  65.473084
            top            71.306308  82.795477  92.098049  53.960273
d           bottom         52.799753  58.670419  51.747686  56.183110
            top            96.873178  76.101363  59.571620  50.971626

In [90]:
#cake_tidy = cake_melt.pivot_table(index = ['recipe_name', 'oven_position', 'temp'])

In [165]:
##cake_tidy.reset_index(inplace = True)
#cake_tidy

3c. Which recipe, on average, is the best? recipe b

In [166]:
#avging accuracy of ea recipe
cake_melt.groupby('recipe_name').mean()

accuracy
recipe_name           
a            63.922201
b            76.736074
c            75.874748
d            62.864844

In [137]:
#pulling the recipe with the highest accuracy
cake_melt.groupby('recipe_name').mean().nlargest(1, 'accuracy')

accuracy
recipe_name           
b            76.736074

In [92]:
cake_tidy.groupby('recipe_name').mean()

accuracy
recipe_name           
a            63.922201
b            76.736074
c            75.874748
d            62.864844

In [95]:
#cake_tidy.groupby('recipe_name').mean().sum(axis=0)

3d. Which oven temperature, on average, produces the best results? 275

In [96]:
cake_melt.head(2)

recipe_name oven_position temp   accuracy
0           a        bottom  225  61.738655
1           a           top  225  51.709751

In [97]:
#avging accuracy for ea temp setting
cake_melt.groupby('temp').mean()

accuracy
temp           
225   71.306022
250   66.577437
275   74.886754
300   66.627655

In [135]:
#then pulling the temp setting with the highest accuracy
cake_melt.groupby('temp').mean().nlargest(1, 'accuracy')

accuracy
temp           
275   74.886754

3e. Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [102]:
cake_tidy = cake_melt.pivot_table(index = ['recipe_name', 'oven_position'], 
                                 columns = 'temp', values = 'accuracy')

In [103]:
cake_tidy

temp                             225        250        275        300
recipe_name oven_position                                            
a           bottom         61.738655  53.912627  74.414730  98.786784
            top            51.709751  52.009735  68.576858  50.228470
b           bottom         57.095320  61.904369  61.196980  99.248541
            top            82.455004  95.224151  98.594881  58.169349
c           bottom         96.470207  52.001358  92.893227  65.473084
            top            71.306308  82.795477  92.098049  53.960273
d           bottom         52.799753  58.670419  51.747686  56.183110
            top            96.873178  76.101363  59.571620  50.971626

In [109]:
cake_tidy.max()

temp
225    96.873178
250    95.224151
275    98.594881
300    99.248541
dtype: float64

In [110]:
cake_tidy.max().max()

99.2485405378462

In [129]:
#group data by accuracy, search for the max
#sorting the data by accuracy
cake = cake_melt.groupby('accuracy').max().sort_values(by='accuracy', ascending=False)

In [132]:
#pull the 1/highest accuracy
cake.head(1)

recipe_name oven_position temp
accuracy                                
99.248541           b        bottom  300

In [133]:
#EVEN FASTER: pulling the highest accuracy
cake_melt.nlargest(1, 'accuracy')

recipe_name oven_position temp   accuracy
26           b        bottom  300  99.248541